driver-core/src/main/com/mongodb/internal/operation/AsyncQueryBatchCursor.java

```diff
@@ -337,7 +338,17 @@ private void endOperationInProgress() {
     private void handleGetMoreQueryResult(final AsyncConnection connection, final SingleResultCallback<List<T>> callback,
                                           final QueryResult<T> result, final boolean tryNext) {
         cursor.set(result.getCursor());
-        if (!tryNext && result.getResults().isEmpty() && result.getCursor() != null) {
+        if (isClosePending) {
+            try {
+                connection.release();
+                if (result.getCursor() == null) {
+                    connectionSource.release();
+                }
+                endOperationInProgress();
```

Comment:
Can we add a test to ensure that this happens and cursors are closed / killed?

Reply:
Seems like this is the spot where we would kill the cursor, if there is still one active.  I don't see that it's being done though.

Reply:
It is the place, and the cursor is killed as a result of the following sequence:
```java
isClosePending == true
    endOperationInProgress()
        close()
            killCursorOnClose()
```
This is exactly the same logic we had (and still have) in the `else` branch of the `handleGetMoreQueryResult` method:
```java
else {
    ...
    else {
        connection.release();
        if (result.getCursor() == null) {
            connectionSource.release();
        }
    }
    endOperationInProgress();
    ...
}
```

Reply:
> Can we add a test to ensure that this happens and cursors are closed / killed?

I thought about adding a test. While I can see how to create a test that checks that the subscription reacts to the cancellation as expected from the standpoint of the enduser, creating a similar test only for the `AsyncQueryBatchCursor` is not as straightforward. I would be glad to discuss how to proceed with such a test. My first thoughts are expressed in [this comment](https://jira.mongodb.org/browse/JAVA-3907?focusedCommentId=3592460&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-3592460) (@rozza, it was written before your request, so we may continue discussing it there).

Reply:
Acknowledged.

Reply:
I think my confusion was due to the fact that below the same connection is used to kill the cursor, while here we release the connection, and check out a new one for the killCursors.  Does that difference matter?

Reply:
We use the provided connection to close the cursor only if we reached the limit, otherwise the code below kills the cursor in a different connection.

The only good reason I can think of for killing the server cursor in a different connection is if the connection we already have is closed or have to be closed due to a failed IO operation in it. I do not understand why the code in `handleGetMoreQueryResult` kills the cursor one way or another, and with regard to killing the cursor, I simply followed the logic that was already in the method.

Reply:
For tests also think of the change stream ticket that is linked. It would be good to confirm that is fixed.